fix(docs-infra): do not break when cookies are disabled in the browser

ajitsinghkaler opened the pull request:
fix(docs-infra): changed window.sessionStorage to default params when cookies disabled

Whenever cookies are disabled windows.sessionStorage is not available for the app so If it throws an error I give it a default value

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
In the current scenario whenever the browser is configured not to accept cookies, the angular.io homepage shows a blank page without any content because windows.sessionStorage is not defined

Issue Number: #33795


## What is the new behavior?
In the new behavior whenever window.sessionStorage is not defined I give it a default value

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ajitsinghkaler commented:
@gkalpak I have issued a pull request can you guide me through the next steps.

gkalpak reviewed (changes requested):
Nice!

The next step would be to add a test for this (something that would fail without your fix and passes with the fix). The unit tests for `ScrollService` are in [aio/src/app/shared/scroll.service.spec.ts][1], so you could add a test somewhere [here][2]:

```ts
it('should not break when cookies are disabled in the browser', () => {
  // Simulate `window.sessionStorage` being inaccessible, when cookies are disabled.
  spyOnProperty(window, 'sessionStorage', 'get').and.throwError('The operation is insecure');

  expect(() => {
    const platformLoc = platformLocation as PlatformLocation;
    const service = new ScrollService(document, platformLoc, viewportScrollerStub, location);

    service.updateScrollLocationHref();
    expect(service.getStoredScrollLocationHref()).toBeNull();

    service.removeStoredScrollInfo();
    expect(service.getStoredScrollPosition()).toBeNull();
  }).not.toThrow();
});
```

As you probably know already, you can run the tests with `yarn test` (from inside the `aio/` directory). Please, make sure that the test fails without your fix and passes with the fix.

[1]: https://github.com/angular/angular/blob/34778f838a/aio/src/app/shared/scroll.service.spec.ts
[2]: https://github.com/angular/angular/blob/34778f838a/aio/src/app/shared/scroll.service.spec.ts#L96

ajitsinghkaler commented:
@gkalpak can you please look into the request and tell me if there is a problem with my code.

gkalpak commented:
The code looks great, @ajitsinghkaler 👍
Now, the final steps are to get CI green and fix the commit messages to more closely follow [our guidelines][1].

1. Rebase your branch on master (this should help get CI green).
2. Squash your commits into one (there should only be one commit, since this is practically one change).
3. Update the commit message (which is already is a pretty good state 💯) to be a little more about what issue is fixed (and less about how we fixed it) and also mention that the commit fixes #33795. I.e. something like:
    > fix(docs-infra): do not break when cookies are disabled in the browser
    >
    > Whenever cookies are disabled in the browser, \`window.sessionStorage\`
    > is not avaialable to the app (i.e. even trying to access
    > \`window.sessionStorage\` throws an error).
    >
    > To avoid breaking the app, we use a no-op \`Storage\` implementation if
    > \`window.sessionStorage\` is not available.
    >
    > Fixes #33795

[1]: https://github.com/angular/angular/blob/a48573efe/CONTRIBUTING.md#commit

mary-poppins commented:
You can preview 7b75cf1 at https://pr33829-7b75cf1.ngbuilds.io/.

mary-poppins commented:
You can preview 4d95ce2 at https://pr33829-4d95ce2.ngbuilds.io/.

mary-poppins commented:
You can preview c9b1619 at https://pr33829-c9b1619.ngbuilds.io/.

mary-poppins commented:
You can preview 3e088f5 at https://pr33829-3e088f5.ngbuilds.io/.

ajitsinghkaler committed:
fix(docs-infra): do not break when cookies are disabled in the browser

Whenever cookies are disabled in the browser, `window.sessionStorage` is not avaialable to the app (i.e. even trying to access `window.sessionStorage` throws an error).

To avoid breaking the app, we use a no-op `Storage` implementation if `window.sessionStorage` is not available.

Fixes #33795

mary-poppins commented:
You can preview 26d5730 at https://pr33829-26d5730.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak now I can't understand why all checks are not passing can you pleas help me

gkalpak commented:
For starters, you need to rebase on master (as mentioned in my previous message).
The error on CI are a little unintuitive at the moment, but we are working on throwing a more descriptive error, when you need to rebase.

(If you are having trouble rebasing on master, lmk and I can do it for you.)

ajitsinghkaler commented:
I've already run git rebase master in my feature branch
Should I rebase my feature branch on the master branch

gkalpak commented:
Thx for trying that, @ajitsinghkaler. `git rebase master` in your feature branch does indeed rebase your branch on your (local) master, so that's should be enough. Maybe your master is not up-to-date, though.

Try the following:

```sh
# Switch to the master branch.
git checkout master

# Fetch changes from `angular/angular` and
# update your local master.
git pull upstream master

# Push the changes to your remote GitHub repo
# (This is not necessary for rebasing your local feature branch,
# but it is a good idea to keep your repo's master up-to-date.)
git push origin master

# Switch to your local feature branch.
git checkout session-storage-bug

# Rebase your feature branch on master.
git rebase master

# Push the updated feature branch to your remote GitHub repo.
# (This will automaticall yupdate the PR.)
git push --force-with-lease
```

(If you get an error about `upstream` not being known, run the following command first: `git remote add upstream https://github.com/angular/angular.git`)

##
Let me know if these commands don't work for you.

mary-poppins commented:
You can preview 6d3b739 at https://pr33829-6d3b739.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak everything is alright I think right now 
Can you please merge the branch

gkalpak reviewed (approved):
LGTM ✨
Thx, @ajitsinghkaler 💯

ajitsinghkaler commented:
@gkalpak No problem. Actually I should thank you for helping me throughout the process.

You can assign me any issue that seems fit for my knowledge. I'll try to help the angular team as much as possible.

gkalpak commented:
That's good to know, @ajitsinghkaler, thx!
Feel free to pick anything from the [hotlist: community-help][1] list.

<sub>(BTW, I would love some help figuring out why there is a visual difference in different OSes in https://github.com/angular/angular/pull/33255#issuecomment-54488570, if you happen to fancy looking into that :wink:)</sub>

[1]: https://github.com/angular/angular/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3A%22hotlist%3A+community-help%22+
